### PR TITLE
Add MPI-enabled unit tests (C/C++ only)

### PR DESCRIPTION
**Pull request by rbberger**
**Summary**

Adds first set of MPI-enabled unit tests. See `c-library/test_library_mpi.cpp`

- add `sublo`, `subhi`, `sublo_lambda`, and `subhi_lambda` to `lammps_extract_global()` and `lammps_extract_global_datatype()`
- add documentation for those flags
- add simple test for global bounding box and subboxes (orthogonal and triclinic)
- test `lammps_gather()` and `lammps_scatter()` methods
- defines a new `add_mpi_test()` CMake function
- add `universe_size`, `universe_rank`, `world_size`, and `world_rank` to `lammps_extract_setting()`
- add tests for above
- add tests for launching LAMMPS through the library interface on a split communicator and in multi-partition mode (2x2 each)
- add support for extracting additional variable styles as char pointer to `lammps_extract_variable()`
- add check for using a "universe" style variable in the multi-partition tests.
- add support for `-in none` so that multi-partition runs can be launched from the library interface without an (initial or dummy) input file.

**Author(s)**

Richard Berger (Temple University) @rbberger 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

Replaces the Google Test default listeners/printer with a custom implementation `MPIPrinter`, defined in the `testing/mpitesting.h` header. Include `test_mpi_main.h` to define a MPI-enabled `main()` method that does the necessary setup.

Rank 0 collects errors from other ranks. That means if an assertion fails on any rank, it will be shown in the output once.
Error messages are will have "[Rank #]` prepended to indicate ownership of the failed assertion.

The `add_mpi_test(name binary nprocs)` CMake function allows adding a test with given `name`, running a `binary` with `nprocs` processors.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Review by sjplimp — APPROVED** [review]

File: unittest/CMakeLists.txt
Comment (junghans) [inline unittest/CMakeLists.txt:5]:
```suggestion
    add_test(NAME ${name} COMMAND ${MPIEXEC_EXECUTABLE} ${MPIEXEC_NUMPROC_FLAG} ${nproc} ${MPIEXEC_PREFLAGS} ${binary} ${MPIEXEC_POSTFLAGS})
```

Newer OpenMPI version disable able oversubscription by default, hence allowing `-DMPIEXEC_PREFLAGS='--oversubscribe'` is sometimes useful. `MPIEXEC_NUMPROC_FLAG` is just a more generic way to write `-np`.

Comment (rbberger) [reply unittest/CMakeLists.txt]: @junghans made some adjustments to make `add_mpi_test` a bit more like `add_test`, adding options like `NAME`, `COMMAND`, `WORKING_DIRECTORY` and `NUM_PROCS`. I also incorporated the `MPIEXEC` variables you mentioned.

**Comment by rbberger** [comment]
@akohlmey waiting for your approval on this one

**Comment by akohlmey** [comment]
> @akohlmey waiting for your approval on this one

This needed a fix when updating to current master. In general, when running on `MPI_COMM_WORLD`, it would be better to use `lammps_open_no_mpi()` (yes, the name is misleading, but not my choice), which implies using the world communicator. `lammps_open()` should be used only when passing a communicator different from `MPI_COMM_WORLD`. Ideally, there would be two sets of tests: one running on all MPI ranks and the other where first the world communicator is split and then tests are run concurrently on a 2:2 split. We should probably also add queries for the values of `Comm::me` and `Comm::nprocs` (and `Comm::nthreads` for completeness) to `lammps_extract_settings()`.

P.S: changing the names of those functions to something more meaningful is on the agenda with some overall changes for a more structured approach to library API names.

**Review by sjplimp — APPROVED** [review]

**Pull request merged**
